testnet/stacks-node/src/tests/signer.rs

```diff
@@ -70,98 +74,545 @@ struct RunningNodes {
-            coordinator_cmd_sender,
-            running_coordinator,
+            signer_cmd_senders,
             running_signers,
             signer_stacks_private_keys,
             stacks_client,
         }
     }
 
+    fn boot_to_epoch_3(&mut self, timeout: Duration) -> Point {
+        boot_to_epoch_3_reward_set(
+            &self.running_nodes.conf,
+            &self.running_nodes.blocks_processed,
+            &self.signer_stacks_private_keys,
+            &self.signer_stacks_private_keys,
+            &mut self.running_nodes.btc_regtest_controller,
+        );
+        let dkg_vote = self.wait_for_dkg(timeout);
+
+        // Advance and mine the DKG key block
+        self.run_until_epoch_3_boundary();
+
+        let reward_cycle = self.get_current_reward_cycle();
+        let set_dkg = self
+            .stacks_client
+            .get_approved_aggregate_key(reward_cycle)
+            .expect("Failed to get approved aggregate key")
+            .expect("No approved aggregate key found");
+        assert_eq!(set_dkg, dkg_vote);
+
+        let (vrfs_submitted, commits_submitted) = (
+            self.running_nodes.vrfs_submitted.clone(),
+            self.running_nodes.commits_submitted.clone(),
+        );
+        // first block wakes up the run loop, wait until a key registration has been submitted.
+        next_block_and(&mut self.running_nodes.btc_regtest_controller, 60, || {
+            let vrf_count = vrfs_submitted.load(Ordering::SeqCst);
+            Ok(vrf_count >= 1)
+        })
+        .unwrap();
+
+        info!("Successfully triggered first block to wake up the miner runloop.");
+        // second block should confirm the VRF register, wait until a block commit is submitted
+        next_block_and(&mut self.running_nodes.btc_regtest_controller, 60, || {
+            let commits_count = commits_submitted.load(Ordering::SeqCst);
+            Ok(commits_count >= 1)
+        })
+        .unwrap();
+        info!("Ready to mine Nakamoto blocks!");
+        set_dkg
+    }
+
+    fn nmb_blocks_to_reward_set_calculation(&mut self) -> u64 {
+        let prepare_phase_len = self
+            .running_nodes
+            .conf
+            .get_burnchain()
+            .pox_constants
+            .prepare_length as u64;
+        let current_block_height = self
+            .running_nodes
+            .btc_regtest_controller
+            .get_headers_height();
+        let curr_reward_cycle = self.get_current_reward_cycle();
+        let next_reward_cycle = curr_reward_cycle.wrapping_add(1);
+        let next_reward_cycle_height = self
+            .running_nodes
+            .btc_regtest_controller
+            .get_burnchain()
+            .reward_cycle_to_block_height(next_reward_cycle);
+        let next_reward_cycle_reward_set_calculation = next_reward_cycle_height
+            .saturating_sub(prepare_phase_len)
+            .wrapping_add(1); // +1 as the reward calculation occurs in the SECOND block of the prepare phase
```

Comment:
Why is this a `wrapping_add` instead of a `saturating_add`?

Reply:
Good question. I don't know what the behaviour for reward cycle is. But I can change this back to saturating add.